### PR TITLE
MONIT-26425-Proxy understands OTLP Cumulative Histograms

### PR DESCRIPTION
**Pull request by deosu**

### CODE REVIEW

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java [hunk 1]
```diff
@@ -340,6 +342,96 @@ public void handlesSummaryAttributes() {
...
+    actualPoints = OtlpProtobufPointUtils.transform(otlpMetric, emptyAttrs, null, DEFAULT_SOURCE);
+
+    assertAllPointsEqual(expectedPoints, actualPoints);
+  }
+
+  @Test
+  public void transformsCumulativeHistogramWithTagLe() {
+    HistogramDataPoint point = HistogramDataPoint.newBuilder()
+        .addAllBucketCounts(ImmutableList.of(1L))
+        .addAttributes(KeyValue.newBuilder().setKey("le").setValue(AnyValue.newBuilder()
```
Comment (oppegard) [inline proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:396]: there's a helper, `OtlpTestHelpers.attribute("le", "someVal")`, to build a `KeyValue` for you.

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java [hunk 2]
```diff
@@ -340,6 +342,96 @@ public void handlesSummaryAttributes() {
...
+
+    assertAllPointsEqual(expectedPoints, actualPoints);
+  }
+
+  @Test
+  public void transformsCumulativeHistogramWithoutBounds() {
+    HistogramDataPoint point = HistogramDataPoint.newBuilder()
+        .addAllBucketCounts(ImmutableList.of(1L))
+        .build();
+    // bucket = 1, datapoint in bucket: 1
```
Comment (oppegard) [inline proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:376]: delete me

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java [hunk 3]
```diff
@@ -340,6 +342,96 @@ public void handlesSummaryAttributes() {
...
+
+  @Test
+  public void transformsCumulativeHistogramWithTagLe() {
+    HistogramDataPoint point = HistogramDataPoint.newBuilder()
+        .addAllBucketCounts(ImmutableList.of(1L))
+        .addAttributes(KeyValue.newBuilder().setKey("le").setValue(AnyValue.newBuilder()
+            .setStringValue("someVal").build()).build())
+        .build();
+    // bucket = 1, datapoint in bucket: 1
+    Histogram histo = Histogram.newBuilder()
```
Comment (oppegard) [inline proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:400]: delete me

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java [hunk 4]
```diff
@@ -340,6 +342,96 @@ public void handlesSummaryAttributes() {
...
+    assertAllPointsEqual(expectedPoints, actualPoints);
+  }
+
+  @Test
+  public void transformsCumulativeHistogramThrowsMalformedDataPointsError() {
+    HistogramDataPoint point = HistogramDataPoint.newBuilder()
+        .addAllExplicitBounds(Collections.singletonList(1.0))
+        .addAllBucketCounts(ImmutableList.of(1L))
+        .build();
+    // bucket = 1, datapoint in bucket: 1
```
Comment (oppegard) [inline proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:422]: delete me

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java [hunk 5]
```diff
@@ -340,6 +342,96 @@ public void handlesSummaryAttributes() {
     }
   }
 
+  @Test
+  public void transformsMinimalCumulativeHistogram() {
+    HistogramDataPoint point = HistogramDataPoint.newBuilder()
+        .addAllExplicitBounds(ImmutableList.of(1.0, 2.0))
+        .addAllBucketCounts(ImmutableList.of(1L, 1L, 1L))
+        .build();
+    // bucket = 1, datapoint in bucket: 1
```
Comment (oppegard) [inline proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:351]: delete me

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 6]
```diff
@@ -162,6 +169,72 @@ private static List<ReportPoint> transformSum(String name, Sum sum,
...
+    switch (histogram.getAggregationTemporality()) {
+      case AGGREGATION_TEMPORALITY_CUMULATIVE:
+        return transformCumulativeHistogram(name, histogram, resourceAttrs);
+//      case AGGREGATION_TEMPORALITY_DELTA:
+//        return transformDeltaHistogram(name, histogram, resourceAttrs);
+//        break;
+      default:
+        return null;
+//        throw new IllegalArgumentException("OTel: histogram with unsupported aggregation " +
+//            "temporality " + histogram.getAggregationTemporality().name());
```
Comment (oppegard) [inline proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:183]: Uncomment me.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 7]
```diff
@@ -162,6 +169,72 @@ private static List<ReportPoint> transformSum(String name, Sum sum,
     return points;
   }
 
+  private static List<ReportPoint> transformHistogram(String name, Histogram histogram, List<KeyValue> resourceAttrs) {
+
```
Comment (oppegard) [inline proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:173]: delete me

**Review by keep94 — CHANGES_REQUESTED** [review]

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java (hunk 7 above)
Comment (deosu) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:173]: done

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 8]
```diff
@@ -115,6 +119,9 @@ public static List<ReportPoint> transform(Metric otlpMetric,
       points.addAll(transformSum(otlpMetric.getName(), otlpMetric.getSum(), resourceAttrs));
     } else if (otlpMetric.hasSummary()) {
       points.addAll(transformSummary(otlpMetric.getName(), otlpMetric.getSummary(), resourceAttrs));
+    } else if (otlpMetric.hasHistogram()) {
```
[the remark this answers is not in the document]
Comment (deosu) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:122]: this was done to avoid setting up a source and preprocessing of rules in each of the methods.

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java (hunk 6 above)
Comment (deosu) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:183]: done

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 9]
```diff
@@ -162,6 +169,72 @@ private static List<ReportPoint> transformSum(String name, Sum sum,
...
+//        throw new IllegalArgumentException("OTel: histogram with unsupported aggregation " +
+//            "temporality " + histogram.getAggregationTemporality().name());
+    }
+  }
+
+  private static List<ReportPoint> transformCumulativeHistogram(String name, Histogram histogram, List<KeyValue> resourceAttrs) {
+
+    List<ReportPoint> reportPoints = new ArrayList<>();
+    for (HistogramDataPoint point : histogram.getDataPointsList()) {
+      reportPoints.addAll(transformHistogramDataPoint(name, point, resourceAttrs));
```
[the remark this answers is not in the document]
Comment (deosu) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:191]: The reason we are not passing the `reportPoints` as an input parameter is to keep all the transform methods consistent. But, it is a good suggestion. If we want to take this suggestion then I think we should update all the transform methods(for Sum, Summary, etc). @oppegard what are your thoughts on this?

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java [hunk 10]
```diff
@@ -162,6 +169,72 @@ private static List<ReportPoint> transformSum(String name, Sum sum,
...
+          "explicit bound counts (" + point.getExplicitBoundsCount() + ") and bucket counts (" +
+          (point.getBucketCountsCount() - 1) + ")");
+    }
+
+    List<ReportPoint> reportPoints = new ArrayList<>();
+    int currentIndex = 0;
+    long cumulativeBucketCount = 0;
+    for (; currentIndex < point.getExplicitBoundsCount(); currentIndex++) {
+      cumulativeBucketCount += point.getBucketCounts(currentIndex);
+      ReportPoint rp = pointWithAnnotations(name, point.getAttributesList(), resourceAttrs,
```
[the remark this answers is not in the document]
Comment (deosu) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:210]: Unit tests are failing post this implementation(reverted the changes). We have to create a new builder every time as the annotations are getting appended after each iteration.

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java (hunk 5 above)
Comment (deosu) [reply proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:351]: done

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java (hunk 2 above)
Comment (deosu) [reply proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:376]: done

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java (hunk 3 above)
Comment (deosu) [reply proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:400]: done

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (oppegard) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:191]: I think this is an optimization that we shouldn't worry about until we have data suggesting we're allocating too many objects. I'm pretty sure the JVM is highly optimized to handle these kinds of situations – it shouldn't be making duplicate copies of the `ReportPoint` objects. There will be a couple wrappers of `List<ReportPoint>` that points to the same `ReportPoint` objects, but that's it.

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java (hunk 1 above)
Comment (deosu) [reply proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:396]: Thank you. Fixed.

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java (hunk 4 above)
Comment (deosu) [reply proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtilsTest.java:422]: done

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (oppegard) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:191]: https://stackoverflow.com/questions/7924887/will-addall-function-in-java-make-a-copy

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java [hunk 11]
```diff
@@ -289,4 +295,63 @@ public void setsSourceFromResourceAttributesNotPointAttributes() {
...
+        .build();
+
+    wavefront.report.ReportPoint wfMetric3 = OtlpTestHelpers.wfReportPointGenerator()
+        .setMetric("test-cumulative-histogram")
+        .setTimestamp(TimeUnit.NANOSECONDS.toMillis(epochTime))
+        .setValue(6)
+        .setHost(DEFAULT_SOURCE)
+        .setAnnotations(Collections.singletonMap("le", "+Inf"))
+        .build();
+
```
Comment (keep94) [inline proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java:343]: Should we test that existing "le" tags get moved to "_le"?

**Review by keep94 — APPROVED** [review]

File: proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (deosu) [reply proxy/src/main/java/com/wavefront/agent/listeners/otlp/OtlpProtobufPointUtils.java:187]: sounds good

File: proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java (hunk 11 above)
Comment (deosu) [reply proxy/src/test/java/com/wavefront/agent/listeners/otlp/OtlpMetricsTest.java:343]: We already have a unit test for this edge case.

**Review by oppegard — APPROVED** [review]
LGTM
